File: be/Security/Permissions.cs
namespace FA23_Convocation2023_API.Security;

public static class Permissions
{
    public const string ManageSystem = "system.manage";
    public const string ManageHalls = "halls.manage";
    public const string ManageSessions = "sessions.manage";
    public const string ManageBachelors = "bachelors.manage";
    public const string CheckIn = "checkin.execute";
    public const string ControlLed = "led.control";
    public const string ManageNotifications = "notifications.manage";
    public const string BroadcastNotifications = "notifications.broadcast";
    public const string ManageMedia = "media.manage";

    public static readonly IReadOnlyDictionary<string, string[]> ByRole =
        new Dictionary<string, string[]>(StringComparer.OrdinalIgnoreCase)
        {
            ["MN"] = new[] { ManageSystem, ManageHalls, ManageSessions, ManageBachelors, CheckIn, ControlLed, ManageNotifications, BroadcastNotifications, ManageMedia },
            ["CK"] = new[] { CheckIn },
            ["MC"] = new[] { ControlLed },
            ["US"] = new[] { ControlLed },
            ["NO"] = new[] { ManageNotifications, BroadcastNotifications }
        };

    public static IReadOnlyCollection<string> ForRoles(IEnumerable<string> roles) =>
        roles.SelectMany(role => ByRole.TryGetValue(role, out var permissions) ? permissions : Array.Empty<string>())
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .OrderBy(permission => permission)
            .ToArray();
}

File: be-nest/src/auth/permissions.ts
export const Permission = {
  ManageSystem: 'system.manage',
  ManageHalls: 'halls.manage',
  ManageSessions: 'sessions.manage',
  ManageBachelors: 'bachelors.manage',
  CheckIn: 'checkin.execute',
  ControlLed: 'led.control',
  ManageNotifications: 'notifications.manage',
  BroadcastNotifications: 'notifications.broadcast',
  ManageMedia: 'media.manage',
} as const;

export type PermissionName = (typeof Permission)[keyof typeof Permission];

export const permissionsByRole: Readonly<Record<string, readonly PermissionName[]>> = {
  MN: Object.values(Permission),
  CK: [Permission.CheckIn],
  MC: [Permission.ControlLed],
  US: [Permission.ControlLed],
  NO: [Permission.ManageNotifications, Permission.BroadcastNotifications],
};

export function expandPermissions(roles: readonly string[]): string[] {
  return [...new Set(roles.flatMap((role) => permissionsByRole[role.toUpperCase()] ?? []))].sort();
}
